circuits/src/lib.rs

```diff
@@ -519,6 +535,25 @@ pub trait MultiProverCircuit<'a, N: 'a + MpcNetwork + Send, S: 'a + SharedValueS
     /// multiplication gate (roughly)
     const BP_GENS_CAPACITY: usize;
 
+    /// Applies all the prover-side constraints embodied by the circuit over
+    /// the allocated witness, secret shares of the counterparty's witness, & statement,
+    /// building them into the given constraint system
+    fn apply_constraints_multi_prover(
+        witness_var: <Self::Witness as CommitSharedProver<N, S>>::SharedVarType,
+        statement: Self::Statement,
+        prover: &mut MpcProver<'a, '_, '_, N, S>,
+        fabric: SharedFabric<N, S>,
+    ) -> Result<(), ProverError>;
+
+    /// Applies all the verifier-side constraints embodied by the circuit over
+    /// the entire committed witness & statement,
+    /// building them into the given constraint system
+    fn apply_constraints_single_prover<CS: RandomizableConstraintSystem>(
+        witness_var: <<Self::WitnessCommitment as Open<N, S>>::OpenOutput as CommitVerifier>::VarType,
+        statement: Self::Statement,
```

Comment:
It's not pretty, but for now I'm just having this method accept a `Self::Statement` parameter. A bit of a leaky abstraction, since the logic this should encapsulate is "given a witness/statement that has been allocated into the constraint system, build the constraints", but now in this method, we first allocate the statement variables and then apply constraints.

The reason for this is that the various statement types don't consistently impl some given trait in a multi-prover context, and given the pending type refactor, I felt it wasn't worthwhile to wrangle all the statement types into a trait implementation.